File: intermediate-package/writing-integ-tests-for-spring-web-apps/rest-api/spring/src/main/java/com/testwithspring/intermediate/user/LoggedInUserService.java
package com.testwithspring.intermediate.user;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class LoggedInUserService implements UserDetailsService {

    private static final Logger LOGGER = LoggerFactory.getLogger(LoggedInUserService.class);

    private final UserRepository repository;

    @Autowired
    LoggedInUserService(UserRepository repository) {
        this.repository = repository;
    }

    @Transactional(readOnly = true)
    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        LOGGER.info("Loading user by username: {}", username);

        User found = repository.findByEmailAddress(username)
                .orElseThrow(() -> new UsernameNotFoundException(String.format(
                        "No user found with username: %s",
                        username
                )));
        LOGGER.info("Found user: {}", found);

        return mapToUserDetailsObject(found);
    }

    private LoggedInUser mapToUserDetailsObject(User user) {
        LoggedInUser userDetails = new LoggedInUser();

        userDetails.setEnabled(user.isEnabled());
        userDetails.setId(user.getId());
        userDetails.setName(user.getName());
        userDetails.setPassword(user.getPassword());
        userDetails.setRole(user.getRole());
        userDetails.setUsername(user.getEmailAddress());

        return userDetails;
    }
}

File: master-package/bonus-lessons-2020/multiple-modules-same-data-sets/integration-tests-rest-api/src/main/kotlin/com/testwithspring/master/user/LoggedInUserService.kt
package com.testwithspring.master.user

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.dao.EmptyResultDataAccessException
import org.springframework.jdbc.core.RowMapper
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Service
import java.sql.ResultSet

/**
 *  Provides a function which queries authentiocated users from the
 *  database.
 */
@Service
open class LoggedInUserService(@Autowired private val jdbcTemplate: NamedParameterJdbcTemplate): UserDetailsService {

    companion object {

        private val LOGGER: Logger = LoggerFactory.getLogger(LoggedInUserService::class.java)

        private const val QUERY_FIND_BY_EMAIL_ADDRESS = "SELECT u.id as id, " +
                "u.email_address AS emailAddress, " +
                "u.is_enabled AS isEnabled, " +
                "u.name AS name, " +
                "u.password AS password, " +
                "u.role AS role " +
                "FROM user_accounts u " +
                "WHERE u.email_address = :emailAddress"

        private fun mapToUserDetailsObject(user: User): LoggedInUser {
            return LoggedInUser(
                    id = user.id,
                    enabled = user.enabled,
                    password = user.password,
                    role = user.role,
                    username = user.emailAddress
            )
        }

        private class UserRowMapper: RowMapper<User> {

            override fun mapRow(resultSet: ResultSet, rowNum: Int): User {
                return User(
                        emailAddress = resultSet.getString("emailAddress"),
                        enabled = resultSet.getBoolean("isEnabled"),
                        id = resultSet.getLong("id"),
                        name = resultSet.getString("name"),
                        password = resultSet.getString("password"),
                        role = UserRole.valueOf(resultSet.getString("role"))
                )
            }
        }
    }

    /**
     * Finds the authenticated user from the database by using the
     * username (aka email address) as a search criteria.
     * @param   username    The username (aka email address) of the authenticated user.
     * @return  The information of the found user.
     * @throws  UsernameNotFoundException   If no user is found with the given username (aka email address).
     */
    override fun loadUserByUsername(username: String): UserDetails {
        LOGGER.info("Find user by username: {}", username)

        try {
            val found = findByUsername(username)
            LOGGER.info("Found user: {}", found)

            return mapToUserDetailsObject(found)
        }
        catch (e: EmptyResultDataAccessException) {
            throw UsernameNotFoundException("No user found with username: $username")
        }
    }

    private fun findByUsername(username: String): User {
        return jdbcTemplate.queryForObject(
                QUERY_FIND_BY_EMAIL_ADDRESS,
                mapOf(Pair("emailAddress", username)),
                UserRowMapper()
        )
    }
}